### ISSUE TITLE
MakeCallback is dangerous, not only or best way of calling callback; provide direct method

### ISSUE THREAD

**Issue by metamatt** [nodejs/nan#284]
NodeJS lets you invoke a "callback" (really, invoke JS code from C++ code) 2 ways:

* calling the callback directly; if you have a `Handle<Function> foo` you can just `foo->Call()`
* calling the callback via `MakeCallback` (in src/node.cc)

The former ("direct") case just calls from C++ into JS synchronously, and returns back into C++ when the JS code returns.

The latter (`MakeCallback`) case calls the JS callback function, but then also ends the current tick, calling `process._tickCallback()` and any *other* callbacks registered via `nextTick()`. This form is intended only for use by C++ code that was dispatched from the libuv event loop, and not for C++ code that was invoked from JS code. Because in that case (JS code "A" running, calls into C++ code in a node addon, the addon's C++ code calls MakeCallback, MakeCallback dispatches the nextTick callbacks) the JS code marked "A" has been preempted, and state changed by the nextTick callbacks can be corrupted. See https://github.com/TooTallNate/node-weak/issues/35 for an example of how dangerous this is.

Use cases for node addons (that would want to use `nan`) wanting to invoke JS code from C++ code that was invoked directly from JS (not from a libuv event callback) include

* node-weak (again, https://github.com/TooTallNate/node-weak/issues/35 is a practical example of this)
* if you want to use V8 interceptors to provide Proxy-like abilities for a dynamic object to trap accesses and call JS handler routines

For these cases, use of `MakeCallback` is not only incorrect but extremely dangerous. The right answer is to call the callback directly. But `nan` doesn't expose a way of doing this; `NanCallback::Call` is actually wrapping `MakeCallback`. So is the right answer to avoid `NanCallback` altogether? I don't think so; it serves a very real need to deal more easily with V8's memory management; using `Persistent<Function>` got harder in newer V8 (node 0.11/0.12) and so it's much nicer to use `NanCallback` instead of raw `Handle<Function>`. So I think `NanCallback` should provide separate methods for calling a callback directly and calling a callback via `MakeCallback`, and make the difference really obvious so people don't accidentally do the latter. (Note in my [patch](https://github.com/TooTallNate/node-weak/pull/36) for the [node-weak instance of this](https://github.com/TooTallNate/node-weak/issues/35), I do use `NanCallback` for memory management, but then have to go to some lengths to extract the real `Handle<Function>` to call directly.)

I think the problem here is exacerbated by (a) nebulous terminology like "callback" (is any function that gets called a "callback"? Not all callbacks are the same) and (b) the difficulty of using `Persistent<Function>` without `NanCallback`. But when you do use `NanCallback` and because of (a), it's just too tempting to accidentally call NanCallback::Call and thus MakeCallback and then preempt your JS code and then suffer all manner of heisenbugs.

So my suggestion and request is:

* add a new `NanCallback::CallDirect` that wraps `Handle<Function>::Call`
* rename `NanCallback::Call` to `NanCallback::CallAndEndTick` (or something similarly explicit that's difficult to confuse)
* in a perfect world, remove `NanCallback::Call`. That will probably be too annoying to addon authors already using it correctly. Then, at least deprecate it (have it generate a compile-time warning, and then act as CallAndEndTick?)

**Comment by kkoopa** [comment]
Hmm, this is actually the first time I've heard of this problem. My experience is completely opposite in that not using `node::MakeCallback` as of 0.11 would lead to a whole bunch of strange errors exactly because `process._tickCallback()` was not called and the same goes for domain support. Normally, every call would want to be done through `node::MakeCallback` for this reason.

Might introduce `NanCallback:: CallDirect` for convenience, but renaming or deprecating `NanCallback::Call` seems less useful. The documentation could however be improved.

**Comment by metamatt** [comment]
See also https://github.com/joyent/node/issues/9245.

**Comment by metamatt** [comment]
@kkoopa, can you elaborate on the "whole bunch of strange errors"?

Barring evidence to the contrary, I'm pretty adamant that inviting people to use `MakeCallback` over-eagerly is a really bad idea. See https://gist.github.com/metamatt/4ce96adbf14de9a97d41, https://gist.github.com/metamatt/99ef2a5e12de2acd5e39. I don't know anything about (haven't tried to use) domain support, so maybe I'm missing something there. But if you call `MakeCallback` in the (I would assume fairly standard/common) case where your addon C++ code was invoked from JS in the same domain, you can preempt the JS code that invoked you and break NodeJS's entire threadless/consistency/concurrency model.

I don't mean to sound too harsh here. But this is not a theoretical concern; this was what actually happened with https://github.com/TooTallNate/node-weak/issues/35 and it caused all manner of heisenbugs in every part of our system that were extremely difficult to track down, and were all caused by this one thing (and all went away when I patched node-weak to not do that).

**Comment by kkoopa** [comment]
In some situations, not using MakeCallback would lead to strange errors such as the next tick never happening, causing all manners of normal code to fail. Now I don't know if this only applied to code invoked from libuv. I do believe this is a genuine concern, but want the common case adjusted for what is normal. It's not exactly nice to have to know, and keep track of, using one way of calling in certain places and other ways in other sections. What I would like is one single way of calling a JS function from C++ that always works, no matter what.

**Comment by metamatt** [comment]
Let's see if we get any traction on joyent/node#9245 then... in https://github.com/TooTallNate/node-weak/pull/36#issuecomment-59133796 @trevnorris said that `MakeCallback` should be safe for use (modulo a small performance hit), or at least that's the intent. I observe otherwise today, but if that becomes true, then we should have your "one single way that always works".

**Comment by kkoopa** [comment]
Yes, let's indeed see what happens with it. I had also been under the impression that MakeCallback should always be safe, which is why it is used everywhere. Having to correctly choose between two different ways of calling is not really acceptable, it will lead to so many hard-to-trace runtime bugs either way one (wrongly) chooses to make the call.

**Comment by saper** [comment]
I actually ran into the problem (https://github.com/sass/node-sass/issues/857) where my application uses uv threadpool to launch worker threads against a C/C++ library. The C++ library invokes certain JS subroutines. In case you are using asynchronous filesystem I/O in those you can end up in a situation where all threadpool threads are blocked (because fs I/O also uses uv threadpool worker threads at the point where non are available/runnable). 

So actually I need to have next ticks invoked since otherwise my application might get stuck forever.
Yes, we are using `NanCallback` but that does not seem to help.

**Comment by kkoopa** [comment]
Nothing can be done for that here as far as I know. `NanCallback` should be the correct choice, as it invokes `node::MakeCallback` under the hood, which handles the `tickCallback`. Perhaps your problem is due to something else? How does it behave under Node 0.10? It might be a bug in libuv or Node for all I know, but I'm not qualified to determine exactly what the reason is.

**Comment by saper** [comment]
Yes, there is nothing nan can do better. We block threads in the uv threadpool; I really wish I could have coroutines there and bility to suspend worker's execution. My solution will probably a separate worker thread communicating with the library. By the way I have an idea for a better fix to TooTallNate/node-weak#35, using `uv_async_t`.

**Comment by hankbao** [comment]
I've just encountered a related issue. We're currently using electron to pack our web app. There's a native node addon providing some functionality. Our web app will call the addon and pass a js callback to it. The native addon store the callback in a NanCallback. It setups a TryCatch before calling the callback. However the TryCatch always fails to catch the exception thrown by the js callback. It turns out that if I call the callback with Function::Call instead of with NanCallback::Call(), the exception can be caught.

I guess my problem may related to this issue.

**Comment by kkoopa** [comment]
MakeCallback does its own error handling.

**Comment by hankbao** [comment]
Unfortunately, when an exception was thrown and caught by MakeCallback in my addon, codes doing cleanup stuffs after calling js callback won't be executed.

**Comment by joshperry** [comment]
This issue and nodejs/node-v0.x-archive#9245 seem to contradict eachother about whether this is an actual bug...

I need to make some synchronous callbacks. If this is an issue, that's fine, however, there are two different ways to call a `Function` depending on the node version and the only place I can find in Nan that glosses over those differences are in `Nan::Callback`. But that calls `MakeCallback`, possibly invoking this bug, but also dispatching the call until after the next tick (besides the overhead of a persistent handle that I don't need).

Does Nan have a primitive to for making synchronous function calls from C++ to JS that takes the new and old methods into account?

**Comment by saper** [comment]
@joshperry how does your C++ code get called? (I mean the code that you'd like to do a sync jump in to JavaScript)? Are you running pure C++ in the v8 event loop? Or is it some code invoked via libuv that runs outside of the v8 event loop (I/O callbacks etc.)?

**Comment by kkoopa** [comment]
Yes, you can always call the v8::Function without using MakeCallback.

**Comment by joshperry** [comment]
I'm working on a usb hardware module that interfaces with libusb. When a js user wants to get a list of devices, I ask libusb for that list, reify it into js object, then (I want to) call the provided callback synchronously. I need it to be synchronous because, after the callback, I need to free the device list.

**Comment by joshperry** [comment]
@kkoopa That's fine, I was just wondering if I needed my own ifdefs to take into account the difference in node <0.12 for the new isolate and context arguments or if Nan had something for me there.

**Comment by kkoopa** [comment]
No isolates are used for calling functions.
```c++
Nan::Callback cb;
*cb->Call();
```
alternatively
```c++
v8::Local<v8::Function> cb;
cb->Call(); 
```

**Comment by joshperry** [comment]
Should I not worry about the deprecation of that `Call` overload? The non-deprecated `Call` requires a context (gotten from the current isolate usually I think) and returns a `Maybe`. Perhaps this is something that Nan will address in the future?

https://v8docs.nodesource.com/io.js-3.0/d4/da0/v8_8h_source.html#l03063

**Comment by saper** [comment]
On Thu, 15 Oct 2015, Joshua Perry wrote:

> Should I not worry about the deprecation of that `Call` overload? The
> non-deprecated `Call` requires a context (gotten from the current
> isolate usually I think) and returns a `Maybe`. Perhaps this is
> something that Nan will address in the future?

If you have a v8::Local<v8::Function> fn, it's no longer a MaybeLocal?

nan has support for checking Maybes and converting them to Locals.

**Comment by joshperry** [comment]
@saper the overload of `Function::Call` that doesn't take a `Context` and doesn't return a `MaybeLocal` is marked `V8_DEPRECATE_SOON("Use maybe version", ...`. The `Call` that isn't to be deprecated does take a context, and does return a `MaybeLocal`.

I don't mind ifdefing this to future proof my own code, but I'd rather use a Nan facility if it exists. Perhaps a feature request issue for a future Nan release @kkoopa?

**Comment by kkoopa** [comment]
You are right that there is no wrapper for directly calling a function. I might add one at some point in the future, but I'd rather see this get fixed in Node so one can always use MakeCallback, as it was supposed to be. I am strongly opposed to the general notion of having a bunch of different ways of calling functions which only work in specific contexts. This is not Perl.

You'll need to do some conditional compilation here.

**Comment by joshperry** [comment]
@kkoopa I agree, but there are some use case (like mine) where an async callback just won't work (Even if this bug is fixed to where MakeCallback is always safe, it is _still_ async).

I _have_ to free these resources, they literally bind up unique hardware resources, and we can't rely on weak persistent callbacks for cleaning such things up: https://code.google.com/p/v8-juice/wiki/CommentaryOnV8#No_GC_guaranty

The conceptual difference between sync and async callbacks is one that is very common when working in pure JS, we need access to that same functionality from V8/Nan.

**Comment by metamatt** [comment]
What do you mean by async callback (presumably going back to "dispatching the call until after the next tick" in https://github.com/nodejs/nan/issues/284#issuecomment-148414280)? In that sentence, did you mean "deferring" instead of "dispatching"? IIRC MakeCallback is synchronous, in that it calls the function you told it to, then does other stuff, then returns; it's the "does other stuff" (which involves ending the current tick, and dispatching all the callbacks associated with that tick) that is problematic in some contexts.

**Comment by trevnorris** [comment]
@kkoopa By "a bunch" I assume you mean 2. Whether the call is made from an existing `HandleScope`, or not. This part is straight forward.

Though even if the issue is fixed it won't address cases outside its scope. Such as whether a `ContextScope` needs to wrap the call.

**Comment by joshperry** [comment]
@metamatt Yes, exactly. I need to call the js function without running the next tick, just like if you were to call a js function from js.

`Nan::Callback` (which uses `node::MakeCallback`) essentially does this:

    function someFunc(callback) {
      var finiteResource = getResource();
      process.nextTick(callback.bind(null, finiteResource));
      // cleanup here? Can't, callback hasn't even run yet.
      cleanupResource(finiteResource);
    }

What I need to do is:

    function someFunc(callback) {
      var finiteResource = getResource();
      callback(finiteResource);
      // I can cleanup here now, because the user code is done
      cleanupResource(finiteResource);
    }

Asking users to call additional functions for freeing memory/resources is just not something that pure-JS people are going to do properly.

**Comment by metamatt** [comment]
Hmm, that's not my memory of `node::MakeCallback`. My day-to-day is on an ancient version of Node (0.12) so maybe things have changed, but glancing at [the current source](https://github.com/nodejs/node/blob/9148114c93861359a502801499d4c26d0b761174/src/async-wrap.cc#L164), I don't think so. `node::MakeCallback` (if my understanding is still correct) is basically a bunch of bookkeeping for domains and exception handling wrapped around: synchronously call the specified callback `cb`, then *additionally* end the current tick which involves calling all the outstanding `nextTick` callbacks associated with the current tick (we end up jumping from C++ back into JS to do that, as the nextTick callback list is maintained in JS code).

If you use `node::MakeCallback`, and thus `Nan::Callback`, to call a JS function from C++, that JS function has run by the time you get back to your C++ code.

**Comment by kkoopa** [comment]
@trevnorris It's a slippery slope, while it might be two for now, the fact that there are more than one could be used as an argument for adding yet another, and so on ad infinitum. Also, while it may seem straightforward to implement, I can guarantee that there will be a steaming pile of never ending bug reports due to someone using the wrong way in one place or another anywhere in a dependency chain, unless this scope business can be strictly enforced at compile time, so that is absolutely impossible to compile an incorrect program.

**Comment by joshperry** [comment]
@metamatt Yes, that is exactly what I _don't_ want. I need the next tick to _not_ run before `Nan::MakeCallback` returns to my C++ code.

Like I said, this distinction (calling a callback with or without deferring) is one that is _very_ frequently used in pure-JS. To properly integrate with the JS world, we have to have that same ability from C++ (which V8 _does_ have, but Nan doesn't have a primitive for version independent calling of non-deferred callbacks)

**Comment by metamatt** [comment]
@joshperry that's the safety issue. I think everyone involved here agrees there needs to be a way for C++ addon code to call a single JS function directly and synchronously and without ending the tick, there's just disagreement on who's responsible for providing a clean API for that, and in the meantime you have to sidestep Nan and use Function::Call directly to achieve that. But it's a safety issue, not a sync vs async issue.

**Comment by metamatt** [comment]
BTW @joshperry I find the terminology around "sync callbacks" and "async callbacks" confusing. I understand the concepts as they're used in JS and agree most JS programmers have at least an intuitive understanding of how these work. But I wouldn't describe them that way. "calling a callback with or without deferring" doesn't really exist; you don't "call a callback with deferring". You tell something else to call your callback later, then you defer, then later happens and your callback gets called (but not by you).

The "sync callback" case is really just function pointers in C/C++, or references to functions in JS, and you can pass them around and tell a function to call you "back" but it all happens now, synchronously, without a return to the event loop.

The "async callback" case is really an event handler, you aren't ever calling the "callback" function directly, you register it with some event dispatch mechanism and say "when event X happens, call function Y".

MakeCallback is synchronous with respect to the callback; it always calls the `cb` argument synchronously. But (because it also ends the tick and dispatches additional callbacks, namely the nextTick callbacks) it's currently only appropriate for use in C++ code that wanted to interpose on the async dispatch path, that is, instead of saying "hey libuv, when event X happens call JS code Y", you said "hey libuv, when event X happens call C++ code Y, and Y knows that it's only invoked from this libuv event handling case that didn't involve the JS event loop, but it ran some JS code that might have called nextTick so now somebody needs to end the tick before returning back to libuv".

**Comment by joshperry** [comment]
@mathiask88 I opened a new issue #495 that is purely focused on my ask. I apologize for my loose terminology and for improperly starting this discussion here.

**Comment by trevnorris** [comment]
@kkoopa I don't see these scenarios as a slippery slope. Either a JS stack has been created or it hasn't. Only v8 could make a change that requires additional circumstances, and I don't think attempting to code today for v8 changes that haven't been announced or even hinted at is worth our time.

From nan's current implementation I see an outstanding issue that nan would have to fix regardless. None of the `MakeCallback()` calls in nan are wrapped in a `HandleScope` or `Context::Scope` (do nan users know these are needed manually?). Both of these are necessary if being called outside the JS stack (how node currently uses `AsyncWrap::MakeCallback`), but neither are necessary if called from within the stack (how node uses `v8::Function::Call`).

**Comment by trevnorris** [comment]
@kkoopa It also may leave users confused as to why specific things have occurred in once scenario but not another. e.g. why the `nextTickQueue` drained before the callback completed, or not.

**Comment by cscott** [comment]
Another option would just be expose something like node's `KickNextTick` (aka, the "second half" of `MakeCallback->Call`), so that "advanced users" can use `Function->Call` (aka "CallDirect"), do their cleanup (or exception-handling), and *then* start the next tick.

For context, I'm working on https://github.com/cscott/node-php-embed, which allows calling into JS from PHP code running in a separate thread.  In my case, calls to async methods from PHP hung the process because the next tick was never triggered when using `Function->Call`; switching to `MakeCallback` [fixed this issue](https://github.com/cscott/node-php-embed/blob/7971f3b0affe56766da01f8cd7786ef1f929952c/src/node_php_jsobject_class.cc#L390) -- but now exceptions are being swallowed.

It would be enough if I could just invoke `Function->Call`, but manually kick the next tick where required.  (As a workaround I'm considering creating a simple no-op function, and then implementing `KickNextTick` by running the no-op function via `MakeCallback`.  I'm going to have to look into the `HandleScope`/`Context::Scope` issue @trevnorris is [talking about](https://github.com/nodejs/nan/issues/284#issuecomment-148453223) though, that's probably going to bite me.

**Comment by kkoopa** [comment]
That scoping issue will be resolved by #499.

**Comment by cscott** [comment]
Should Nan::AsyncWorker be entering a `Context::Scope` in its `WorkComplete` method?  It is eventually invoking `Nan::Callback`, which as I understand this discussion, eventually enters a `Context::Scope` before calling back into JS -- but it seems like we're entering this scope "too late" -- presumably subclasses might want to do some JS work in an override of `WorkComplete` before invoking the callback (perhaps just marshalling arguments), and this should be done in the appropriate context.

That is, I'm wondering if the "enter a context scope" part of MakeCallback really ought to be factored out into something more useful for JS code in general which is invoked from the uv event loop.  That's the part of `MakeCallback` which is not performed by `Function->Call` and does not involve kicking the next tick.

**Comment by kkoopa** [comment]
Good question. `Nan::Callback` uses `node::MakeCallback` to execute, which does seem to enter a `Context::Scope` nowadays. It did not do so in the past, say 0.8.28, possibly because `Context::Scope` did not exist.

I really don't know what a `Context::Scope` is supposed to do in Node, as it is not a browser, and the Embedder's Guide says it's meant for running many embedded scripts without them interfering with each other. How does that translate to Node? Why not simply use a single `Context` everywhere and be satisfied with it?

I have not seen much documentation related to any of this in the first place. Hopefully we could get it sorted out in a nice way so that it can be documented or hidden away in NAN so nobody else needs to dig through this.

**Comment by saper** [comment]
Isn't `Context::Scope` about keeping tab on `Local<T>` variables and purging them?

I think if you don't create any own `Local`s in the `HandleOKCallback` callback you don't have to create your own scope there. Actually `WorkComplete` creates the scope as the safety net for user-supplied `HandleOKCallback()` and `HandleErrorCallback()`. So, if you for some reason need your own `WorkComplete` method - you need your scope if you play with `Local`s. Otherwise your scope will  be whatever was the last one in the current V8 isolate, which may cause those variables to live longer than they should.

**Comment by kkoopa** [comment]
But that should be what `HandleScope` is for... Is `HandleScope` independent of `Context::Scope`?
https://developers.google.com/v8/embed#handles
https://developers.google.com/v8/embed#contexts

Here it shows both `Context::Scope` and `HandleScope`
[Handles](https://developers.google.com/v8/images/local_persist_handles_review.png)

Here it shows using `Context::Scope` to enter a context.
[Contexts](https://developers.google.com/v8/images/intro_contexts.png)

**Comment by saper** [comment]
My mistake. I think we shouldn't do anything with Context Scopes at all? Looks like node provides one for us. Regarding independence: `Context::Scope` seems to be just like a `Local<>`, `CheckImmediate()` in `src/node.cc` establishes a handle scope first and then creates a `Context::Scope` local variable.

**Comment by trevnorris** [comment]
> I think if you don't create any own Locals in the HandleOKCallback callback you don't have to create your own scope there.

You do automatically by running `v8::Function::Call()`. Also if domains are in use then another `v8::Object` will be created in `MakeCallback` as well.

@kkoopa May consider implementing wrapping `WorkComplete` with a `v8::SealHandleScope` (I have a backported patch for v0.10 if needed) to help devs ensure they don't accidentally leak any handles.

As for the `Context::Scope` I'm investigating that. node has one before every entry to `MakeCallback`, but the git log does not make it obvious why.

**Comment by kkoopa** [comment]
@trevnorris Oh, yes, please submit backported `SealHandleScope` (I presume it can be made to work on 0.8 as well). It could also be used in all the test code to ensure that NAN functions don't leak.

**Comment by bnoordhuis** [comment]
> As for the Context::Scope I'm investigating that. node has one before every entry to MakeCallback, but the git log does not make it obvious why.

That was me.  It lets multiple contexts share a single event loop.  nw.js is a prime example.

**Comment by trevnorris** [comment]
@kkoopa The patch is pretty small, but does require a patch to v8. Not sure if it's 100% correct, but has worked well enough for my testing. Not sure if it can work on v0.8. Here's the patch: https://gist.github.com/trevnorris/4802d21ba3044a0c30b0

**Comment by kkoopa** [comment]
Aww, patching V8 is not really possible, save for internal leak testing. Maybe it is possible to hack something up by wrapping a `v8::HandleScope` in a `Nan::SealHandleScope` and doing the checking by hand. I'll take a look, but don't feel optimistic. I really wish C++ had a proper way of overriding access modifiers.

**Comment by trevnorris** [comment]
@kkoopa sorry, the patch is actually only for v0.12. don't have a working one for v0.10. I can submit it to land, but w/o v0.10 and v0.8 support I don't think it'll be worth it.

**Comment by kkoopa** [comment]
Nah, not here and now, but do submit it for 0.12 anyway, hopefully we will get to drop 0.8 and 0.10 in 12-18 months (and start forcing Isolate passing everywhere and other good stuff).

**Comment by bnoordhuis** [comment]
Apropos v0.8, what's the reason nan still supports it?  Upstream no longer does.  I'd just sunset the support and say "nan version $x is the last one to support v0.8."

**Comment by kkoopa** [comment]
It's so similar to 0.10 that it is trivial to support. Npm also support 0.8 for a while more. My plan was to drop both 0.8 and 0.10 at the same time, since nothing is gained by dropping only 0.8.

**Comment by bnoordhuis** [comment]
It would give people an incentive to upgrade to a version that's still supported by upstream.  Although if you're still stuck on v0.8 at this point, you're probably not a very upgrade-y person anyway.

**Comment by kkoopa** [comment]
Yeah, but it is not so much for end-users as for developers. Their modules can run on 0.8, even if the majority of their users couldn't care less. That makes upgrading their NAN version a non-issue, giving support for new versions at no expense.

From NAN's perspective there are three versions at the moment: 0.8 to 0.10, 0.12 to io.js 2, and io.js 3 to Node 5+. These correspond to: ancient, isolates, maybes.

**Comment by LinusU** [comment]
I got bit by this today as well, I would love to see `NanCallback::CallDirect` implemented. Would a pull request be accepted?

I'm also in favor for renaming `Call` to something more explicit, but that's step 2. I think we should add CallDirect either way...

**Comment by kkoopa** [comment]
I don't think adding that is the right approach. Just get the underlying function and use Nan::Call on it once that has been added from another PR.

Nan::Call(*callback, Nan::GetCurrentContext()->Global(), 0, 0);

**Comment by LinusU** [comment]
@kkoopa Perfect, that's exactly what I need!

**Comment by kkoopa** [comment]
For reference, `Nan:Call` is #496, which is merely a wrapper for `v8::Function::Call`. `Nan::Callback` already supports the `*` operator as a shorthand for `Nan::Callback::GetFunction`.

**Comment by kkoopa** [comment]
I'll close this, as it should no longer be an issue. `Nan::Call` is included in 2.2 and `node::MakeCallback` is now reentrant.
